File: backend/src/main/java/com/example/backend/entities/Buying.java
package com.example.backend.entities;

import com.example.backend.dtos.in.buyings.CreateBuyingDto;
import com.example.backend.types.BuyingCategory;
import jakarta.persistence.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;

@Entity
@Table(name = "buying")
public class Buying {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "apartment_id", nullable = false)
    private Apartment apartment;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "created_by", nullable = false)
    private Profile createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to")
    private Profile assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "completed_by")
    private Profile completedBy;

    @Column(name = "name", nullable = false, length = 64)
    private String name;

    @Column(name = "quantity", nullable = false, length = 16)
    private String quantity;

    @Enumerated
    @Column(name = "category", nullable = false)
    private BuyingCategory  category = BuyingCategory.OTHER;

    @Column(name = "is_public", nullable = false)
    private Boolean isPublic = false;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "completed_at")
    private Instant completedAt;

    public Buying() {}

    public Buying(
            Profile createdBy,
            Profile assignedTo,
            Apartment apartment,
            String name,
            String quantity,
            BuyingCategory category,
            Boolean isPublic
    ) {
        this.createdBy = createdBy;
        this.assignedTo = assignedTo;
        this.apartment = apartment;
        this.name = name;
        this.quantity = quantity;
        if (category != null)
            this.category = category;
        this.isPublic = isPublic;
    }

    public Instant getCompletedAt() {
        return completedAt;
    }

    public void setCompletedAt(Instant completedAt) {
        this.completedAt = completedAt;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Boolean getIsPublic() {
        return isPublic;
    }

    public void setIsPublic(Boolean isPublic) {
        this.isPublic = isPublic;
    }

    public BuyingCategory getCategory() {
        return category;
    }

    public void setCategory(BuyingCategory category) {
        this.category = category;
    }

    public String getQuantity() {
        return quantity;
    }

    public void setQuantity(String quantity) {
        this.quantity = quantity;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Profile getCompletedBy() {
        return completedBy;
    }

    public void setCompletedBy(Profile completedBy) {
        this.completedBy = completedBy;
    }

    public Profile getAssignedTo() {
        return assignedTo;
    }

    public void setAssignedTo(Profile assignedTo) {
        this.assignedTo = assignedTo;
    }

    public Profile getCreatedBy() {
        return createdBy;
    }

    public Apartment getApartment() {
        return apartment;
    }

    public Long getId() {
        return id;
    }


}

File: frontend/lib/shared/data/models/buying.dart
import 'package:frontend/shared/data/models/profile.dart';
import 'package:frontend/shared/data/types/buying_category.dart';
import 'package:frontend/shared/utils/util_functions.dart';

class Buying {
  Buying({
    required this.id,
    required this.createdBy,
    this.assignedTo,
    this.completedBy,
    required this.name,
    required this.quantity,
    required this.category,
  });

  factory Buying.fromJson(Map<String, dynamic> json) {
    return Buying(
      id: json["id"] as int,
      createdBy: Profile.fromJson(json["createdBy"]),
      assignedTo: json["assignedTo"] != null
          ? Profile.fromJson(json["assignedTo"])
          : null,
      completedBy: json["completedBy"] != null
          ? Profile.fromJson(json["completedBy"])
          : null,
      name: json["name"] as String,
      quantity: json["quantity"] as String,
      category: UtilFunctions.getTValueFromName<BuyingCategory>(
        json["category"],
      ),
    );
  }

  final int id;
  final Profile createdBy;
  final Profile? assignedTo;
  final Profile? completedBy;
  final String name;
  final String quantity;
  final BuyingCategory category;
}
